client/rest/src/main/java/org/elasticsearch/client/RestClient.java

```diff
@@ -361,11 +361,19 @@ public void cancelled() {
     private void setHeaders(HttpRequest httpRequest, Header[] requestHeaders) {
         Objects.requireNonNull(requestHeaders, "request headers must not be null");
         for (Header defaultHeader : defaultHeaders) {
-            httpRequest.setHeader(defaultHeader);
+            httpRequest.addHeader(defaultHeader);
         }
-        for (Header requestHeader : requestHeaders) {
-            Objects.requireNonNull(requestHeader, "request header must not be null");
-            httpRequest.setHeader(requestHeader);
+        if (requestHeaders.length != 0) {
+            final Set<String> uniqueNames = new HashSet<>();
+            for (Header requestHeader : requestHeaders) {
+                Objects.requireNonNull(requestHeader, "request header must not be null");
+                // overwrite any default header
+                if (uniqueNames.add(requestHeader.getName())) {
+                    httpRequest.setHeader(requestHeader);
+                } else {
+                    httpRequest.addHeader(requestHeader);
```

Comment:
I flipped it around and added a comment:

1. Add all request headers.
2. Add any default header that was not a request header.

This still uses the request-level `Set`, but it avoids ever calling `setHeader` in favor of always using `addHeader`. It definitely simplifies it.